packages/sol-resolver/src/resolvers/npm_resolver.ts

```diff
@@ -1,42 +1,86 @@
+            ? packageName
+            : path.join(packageScopeIfExists, packageName);
+        const internalPath = path.join(...other);
+        return {
+            dependencyName,
+            internalPath,
+        };
+    }
+    private static _getPackageVersion(packagePath: string): string {
+        const packageJSONPath = path.join(packagePath, 'package.json');
+        const packageJSON = JSON.parse(fs.readFileSync(packageJSONPath).toString());
+        const packageVersion = packageJSON.version;
+        return packageVersion;
+    }
     constructor(packagePath: string) {
         super();
         this._packagePath = packagePath;
+        const packageJSONPath = path.join(this._packagePath, 'package.json');
+        const packageJSON = JSON.parse(fs.readFileSync(packageJSONPath).toString());
+        this._dependencies = { ...packageJSON.dependencies, ...packageJSON.devDependencies };
     }
     public resolveIfExists(importPath: string): ContractSource | undefined {
-        if (!importPath.startsWith('/')) {
-            let packageName;
-            let packageScopeIfExists;
-            let other;
-            if (_.startsWith(importPath, '@')) {
-                [packageScopeIfExists, packageName, ...other] = importPath.split('/');
-            } else {
-                [packageName, ...other] = importPath.split('/');
-            }
-            const pathWithinPackage = path.join(...other);
-            let currentPath = this._packagePath;
-            const ROOT_PATH = '/';
-            while (currentPath !== ROOT_PATH) {
-                const packagePath = _.isUndefined(packageScopeIfExists)
-                    ? packageName
-                    : path.join(packageScopeIfExists, packageName);
-                const lookupPath = path.join(currentPath, 'node_modules', packagePath, pathWithinPackage);
-                if (fs.existsSync(lookupPath) && fs.lstatSync(lookupPath).isFile()) {
-                    const fileContent = fs.readFileSync(lookupPath).toString();
-                    return { source: fileContent, path: importPath, absolutePath: lookupPath };
-                }
-                currentPath = path.dirname(currentPath);
-            }
+        if (importPath.startsWith('/')) {
+            return undefined;
         }
-        return undefined;
+        const { dependencyName, internalPath } = NPMResolver._parseDependencyImport(importPath);
+        const dependencyLookupTarget = path.join('node_modules', dependencyName, internalPath);
+        const dependencyLookupResult = NPMResolver._lookupPath(this._packagePath, dependencyLookupTarget);
+        if (_.isUndefined(dependencyLookupResult)) {
+            return undefined;
+        }
+        if (_.isUndefined(this._dependencies[dependencyName])) {
+            throw new Error(
+                `Lookup for ${dependencyName} started from ${this._packagePath} failed. Not specified in package.json`,
+            );
+        }
```

Comment:
Move this check up after the call to `NPMResolver._parseDependencyImport`. It does not depend on looking up the path and we should fail as fast as possible.